### PR TITLE
Fix Unix stack overflow trace logging for native failure locations

### LINKED ISSUE #32203: Create CoreCLR tests for verifying the stack overflow stack trace printing
The tests should verify that at stack overflow caused by both probing and real stack overflow from one or more threads at the same time prints expected stack trace.

### LINKED ISSUE #35420: Stack overflow reporting on Unix doesn't work when failing in native runtime
While testing a stack overflow handling change on OSX, I have discovered that when stack overflow happens in native runtime code (QCALL, FCALL, ...), we don't try to invoke the handler in the runtime since we consider the location unsafe. The result is that the process is terminated with SIGSEGV without any message.

I wonder if for stack overflow, we could be less strict in what location we would consider safe for handling this hardware exception and risk that in rare cases, we could hit a deadlock.
But we should print the stack overflow message again, at least.

### PR DESCRIPTION
**Pull request by janvorli**
When stack overflow happens in native runtime code (e.g. in one of the
QCALLs, FCALLs etc.), we do not print stack trace to the console
and we fail fast with just a "Stack overflow" message. The reason is
that the stack overflow handling is executed on the failing thread and
it is not safe to call methods that are not async safe (reentrant).
The reason is that we could end up hanging or crashing when trying to
report the stack overflow.

However, we have the same potential issue on Windows and yet we try to
report the stack trace anyways. So this change modifies the behavior on
Unix to be the same.

I have also added a bunch of stack overflow reporting tests.

Close #35420
Close #32203

### CODE REVIEW

File: src/coreclr/tests/src/baseservices/exceptions/stackoverflow/stackoverflow.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
+using System;
```
Comment (jkotas) [inline src/coreclr/tests/src/baseservices/exceptions/stackoverflow/stackoverflow.cs:1]: Nit: License headers

**Review by jkotas — APPROVED** [review]

File: src/coreclr/tests/src/baseservices/exceptions/stackoverflow/stackoverflow.cs (hunk 1 above)
Comment (janvorli) [reply src/coreclr/tests/src/baseservices/exceptions/stackoverflow/stackoverflow.cs:4]: Oops, thank you for noticing that, fixed.

**Comment by janvorli** [comment]
FYI: @danmosemsft

**Comment by danmosemsft** [comment]
Thanks for doing this @janvorli . This is definitely valuable (just today I was looking at a SO issue reported)

**Pull request merged**